File: hello-go/src/client/core/download_manager.go
package core

import (
	"fmt"

	"github.com/feuyeux/hello-mmap/hello-go/src/logger"
)

func Download(ws *WebSocketClient, streamID string, outputPath string, fileSize int64) error {
	var offset int64 = 0
	var bytesReceived int64 = 0
	lastProgress := 0
	isFirstChunk := true

	for offset < fileSize {
		// Calculate how much data we still need
		remainingBytes := fileSize - offset
		chunkSize := int(Min(int64(ChunkSize), remainingBytes))

		// Send GET message
		offsetPtr := offset
		lengthPtr := chunkSize
		logger.Debug(fmt.Sprintf("Requesting chunk at offset %d, length %d (remaining: %d)", offset, chunkSize, remainingBytes))
		err := ws.SendControlMessage(ControlMessage{
			Type:     "GET",
			StreamID: streamID,
			Offset:   &offsetPtr,
			Length:   &lengthPtr,
		})
		if err != nil {
			return fmt.Errorf("failed to send GET message: %w", err)
		}

		// Receive binary data - one GET request = one binary response
		// The server may send less data than requested
		logger.Debug(fmt.Sprintf("Waiting for binary data at offset %d", offset))
		data, err := ws.ReceiveBinary()
		if err != nil {
			return fmt.Errorf("failed to receive data: %w", err)
		}

		logger.Debug(fmt.Sprintf("Received %d bytes of data", len(data)))

		if len(data) == 0 {
			return fmt.Errorf("no data received for offset %d", offset)
		}

		// Write to file
		if err := WriteChunk(outputPath, data, !isFirstChunk); err != nil {
			return fmt.Errorf("failed to write chunk: %w", err)
		}

		isFirstChunk = false
		offset += int64(len(data))
		bytesReceived += int64(len(data))

		// Report progress
		progress := int(bytesReceived * 100 / fileSize)
		if progress >= lastProgress+25 && progress <= 100 {
			logger.Info(fmt.Sprintf("Download progress: %d/%d bytes (%d%%)", bytesReceived, fileSize, progress))
			lastProgress = progress
		}
	}

	// Ensure 100% is reported
	if lastProgress < 100 {
		logger.Info(fmt.Sprintf("Download progress: %d/%d bytes (100%%)", fileSize, fileSize))
	}

	return nil
}

File: hello-rust/src/client/download_manager.rs
use super::{
    file_manager,
    websocket_client::{ControlMessage, WebSocketClient},
};
use crate::logger;
use anyhow::{Context, Result};

pub async fn download(
    ws_client: &mut WebSocketClient,
    stream_id: &str,
    output_path: &str,
    file_size: u64,
) -> Result<()> {
    let mut offset = 0u64;
    let mut bytes_received = 0u64;
    let mut last_progress = 0;
    let mut is_first_chunk = true;

    while offset < file_size {
        let chunk_size =
            std::cmp::min(file_manager::CHUNK_SIZE as u64, file_size - offset) as usize;

        // Send GET message
        let get_msg = ControlMessage {
            msg_type: "GET".to_string(),
            stream_id: Some(stream_id.to_string()),
            offset: Some(offset),
            length: Some(chunk_size),
            message: None,
        };
        ws_client.send_control_message(get_msg).await?;

        // Receive binary data
        let data = ws_client.receive_binary().await?;

        // Write to file
        file_manager::write_chunk(output_path, &data, !is_first_chunk)
            .await
            .context("Failed to write downloaded chunk")?;

        is_first_chunk = false;
        offset += data.len() as u64;
        bytes_received += data.len() as u64;

        // Report progress
        let progress = (bytes_received * 100 / file_size) as usize;
        if progress >= last_progress + 25 && progress <= 100 {
            logger::log_info(&format!(
                "Download progress: {}/{} bytes ({}%)",
                bytes_received, file_size, progress
            ));
            last_progress = progress;
        }
    }

    // Ensure 100% is reported
    if last_progress < 100 {
        logger::log_info(&format!(
            "Download progress: {}/{} bytes (100%)",
            file_size, file_size
        ));
    }

    Ok(())
}
